js/plugins/google-genai/src/googleai/gemini.ts

```diff
@@ -0,0 +1,794 @@
+      }
+
+      const msg = toGeminiMessage(messages[messages.length - 1], model);
+
+      let generateContentRequest: GenerateContentRequest = {
+        systemInstruction,
+        generationConfig,
+        tools: tools.length ? tools : undefined,
+        toolConfig,
+        safetySettings: safetySettingsFromConfig?.filter(
+          (setting) => setting.category !== 'HARM_CATEGORY_UNSPECIFIED'
+        ) as SafetySetting[],
+        cachedContent: undefined,
+        contents: messages.map((message) => toGeminiMessage(message, model)),
+      };
+
+      const modelVersion = (versionFromConfig ||
+        model.version ||
+        apiModelName) as string;
+
+      apiKey = apiKeyFromConfig || apiKey;
+      if (!apiKey) {
+        throw new GenkitError({
+          status: 'INVALID_ARGUMENT',
+          message:
+            'GoogleAI plugin was initialized with {apiKey: false} but no apiKey configuration was passed at call time.',
+        });
+      }
+
+      const callGemini = async () => {
+        let response: GenerateContentResponse;
+
+        if (sendChunk) {
+          const result = await generateContentStream(
+            apiKey!,
+            modelVersion,
+            generateContentRequest,
+            options
+          );
+
+          for await (const item of result.stream) {
+            item.candidates?.forEach((candidate) => {
+              const c = fromGeminiCandidate(candidate);
+              sendChunk({
+                index: c.index,
+                content: c.message.content,
+              });
+            });
+          }
+          response = await result.response;
+        } else {
+          response = await generateContent(
+            apiKey!,
+            modelVersion,
+            generateContentRequest,
+            options
+          );
+        }
+
+        const candidates = response.candidates || [];
+        if (response.candidates?.['undefined']) {
+          candidates.push(response.candidates['undefined']);
+        }
+        if (!candidates.length) {
+          throw new GenkitError({
+            status: 'FAILED_PRECONDITION',
+            message: 'No valid candidates returned.',
+          });
+        }
+
+        const candidateData = candidates.map(fromGeminiCandidate) || [];
+
+        return {
+          candidates: candidateData,
+          custom: response,
+          usage: {
+            ...getBasicUsageStats(request.messages, candidateData),
+            inputTokens: response.usageMetadata?.promptTokenCount,
+            outputTokens: response.usageMetadata?.candidatesTokenCount,
+            totalTokens: response.usageMetadata?.totalTokenCount,
+            cachedContentTokens:
+              response.usageMetadata?.cachedContentTokenCount,
+          },
+        };
+      };
+
+      // If debugTraces is enabled, we wrap the actual model call with a span, add raw
+      // API params as for input.
+      return debugTraces
+        ? await runInNewSpan(
```

Comment:
please move the debug trace into `import {  generateContent,  generateContentStream } from './client';`
it should be shared logic

Reply:
Discussed offline. To be done at a later date.